quintilesims/layer0
# Terraform update

**What does this pull request do?**
The main point of this PR is to update the terraform plugin. I've updated all of the code to work with the update client / job model. I've also slightly changed the unit test patterns for this package. While testing this, I encountered quite a few bugs/missing features needed to properly test these changes. With that, I've also fixed the following bugs / added the missing functionality: 

* Adds tags for jobs
* Adds a custom error handler to the api, so all uncaught errors are still returned as a `ServerError`. It also maps our error codes to meaningful http status codes. 
* Fixes cli text printer tests and implementation for tasks
* Adds retry logic to service create, since that would sometimes fail with the "failed to assume role and validate ports" error, due to the loadbalancer still coming up. 
* Adds an optional `scale` param to service create 
* Adds the missing job command to the cli 
* Adds shorthand flag names for `output` and `timeout`, e.g. `l0 -o json ...`

**How should this be tested?**
Unit test all of the package where I have made changes. 
To test the terraform provider, I usually just create a `main.tf` with every available resource and data source. I add/update/remove each of the resources' options until i've exhausted all possible configurations. You also need to test that it gracefully handles when things no longer exist - e.g. it was created in terraform but deleted by the CLI. 


**Checklist**
